File: examples/typescript/symbio-kb-client.ts
/**
 * SymbioKnowledgeBase Agent API — TypeScript Client
 *
 * A production-ready client for the SymbioKnowledgeBase Agent API.
 * Provides typed access to pages, search, and knowledge graph endpoints
 * with built-in rate limit handling and retry logic.
 *
 * @example
 * ```ts
 * import { SymbioKB } from "./symbio-kb-client.js";
 *
 * const kb = new SymbioKB({
 *   baseUrl: "https://kb.example.com",
 *   apiKey: "skb_live_...",
 * });
 *
 * const results = await kb.search("deployment guide");
 * const page = await kb.readPage(results.results[0].page_id);
 * ```
 */

// ---------------------------------------------------------------------------
// Interfaces
// ---------------------------------------------------------------------------

/** A knowledge base page. */
export interface Page {
  id: string;
  title: string;
  icon?: string | null;
  parent_id?: string | null;
  markdown?: string;
  created_at?: string;
  updated_at?: string;
}

/** A single full-text search result. */
export interface SearchResult {
  page_id: string;
  title: string;
  snippet: string;
  score: number;
  icon?: string | null;
}

/** Standard API response envelope for single items. */
export interface ApiResponse<T> {
  data: T;
  meta: {
    timestamp: string;
    [key: string]: unknown;
  };
}

/** Standard API response envelope for lists with pagination. */
export interface ApiListResponse<T> {
  data: T[];
  meta: {
    total: number;
    limit: number;
    offset: number;
    timestamp: string;
  };
}

/** Standard API error envelope. */
export interface ApiErrorBody {
  error: {
    code: string;
    message: string;
    details?: unknown[];
  };
  meta: {
    timestamp: string;
  };
}

/** Options for constructing a SymbioKB client. */
export interface SymbioKBOptions {
  /** Base URL of the SymbioKnowledgeBase instance. */
  baseUrl: string;
  /** Bearer token: API key (skb_live_*) or Supabase JWT. */
  apiKey: string;
  /** HTTP request timeout in milliseconds. Default: 30000. */
  timeout?: number;
  /** Maximum automatic retries on 429 responses. Default: 3. */
  maxRetries?: number;
}

/** Paginated page list result. */
export interface PageListResult {
  pages: Page[];
  total: number;
  limit: number;
  offset: number;
}

/** Paginated search result list. */
export interface SearchResults {
  results: SearchResult[];
  total: number;
  limit: number;
  offset: number;
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

/** Base error class for all SymbioKB client errors. */
export class SymbioKBError extends Error {
  code?: string;
  status?: number;

  constructor(message: string, code?: string, status?: number) {
    super(message);
    this.name = "SymbioKBError";
    this.code = code;
    this.status = status;
  }
}

/** Raised when the API key is missing, invalid, or revoked (HTTP 401). */
export class AuthenticationError extends SymbioKBError {
  constructor(message: string, code?: string) {
    super(message, code, 401);
    this.name = "AuthenticationError";
  }
}

/** Raised when the API key lacks the required scope (HTTP 403). */
export class ForbiddenError extends SymbioKBError {
  constructor(message: string, code?: string) {
    super(message, code, 403);
    this.name = "ForbiddenError";
  }
}

/** Raised when the requested resource does not exist (HTTP 404). */
export class NotFoundError extends SymbioKBError {
  constructor(message: string, code?: string) {
    super(message, code, 404);
    this.name = "NotFoundError";
  }
}

/** Raised when request parameters fail server-side validation (HTTP 400). */
export class ValidationError extends SymbioKBError {
  constructor(message: string, code?: string) {
    super(message, code, 400);
    this.name = "ValidationError";
  }
}

/** Raised when rate limiting is exceeded and all retries are exhausted. */
export class RateLimitError extends SymbioKBError {
  retryAfter?: number;

  constructor(message: string, retryAfter?: number, code?: string) {
    super(message, code, 429);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

/** Raised for unexpected server-side errors (HTTP 5xx). */
export class ServerError extends SymbioKBError {
  constructor(message: string, code?: string, status?: number) {
    super(message, code, status ?? 500);
    this.name = "ServerError";
  }
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/**
 * Client for the SymbioKnowledgeBase Agent API.
 *
 * All methods are async and return typed results. Rate limiting (HTTP 429)
 * is handled automatically with exponential back-off guided by the
 * `Retry-After` header.
 */
export class SymbioKB {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly timeout: number;
  private readonly maxRetries: number;

  constructor(options: SymbioKBOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.timeout = options.timeout ?? 30_000;
    this.maxRetries = options.maxRetries ?? 3;
  }

  // -----------------------------------------------------------------------
  // Internal helpers
  // -----------------------------------------------------------------------

  private url(path: string): string {
    return `${this.baseUrl}/api/agent${path}`;
  }

  /** Sleep for the given number of milliseconds. */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Execute an HTTP request with automatic rate-limit retry logic.
   */
  private async request<T>(
    method: string,
    path: string,
    options?: {
      params?: Record<string, string | number>;
      body?: Record<string, unknown>;
    },
  ): Promise<T> {
    const url = new URL(this.url(path));

    if (options?.params) {
      for (const [key, value] of Object.entries(options.params)) {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      let response: Response;
      try {
        response = await fetch(url.toString(), {
          method,
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "Content-Type": "application/json",
            Accept: "application/json",
          },
          body: options?.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });
      } catch (err) {
        clearTimeout(timeoutId);
        if (err instanceof DOMException && err.name === "AbortError") {
          throw new SymbioKBError(`Request timed out after ${this.timeout}ms`);
        }
        throw new SymbioKBError(
          `HTTP request failed: ${err instanceof Error ? err.message : String(err)}`,
        );
      } finally {
        clearTimeout(timeoutId);
      }

      // --- Rate limit handling (429) ---
      if (response.status === 429) {
        const retryAfter = parseInt(response.headers.get("Retry-After") ?? "5", 10);
        if (attempt < this.maxRetries) {
          await this.sleep(retryAfter * 1000);
          continue;
        }
        throw new RateLimitError(
          "Rate limit exceeded after all retries",
          retryAfter,
          "RATE_LIMIT_EXCEEDED",
        );
      }

      // --- Parse response body ---
      let body: unknown;
      try {
        body = await response.json();
      } catch {
        throw new ServerError(
          `Non-JSON response (HTTP ${response.status})`,
          undefined,
          response.status,
        );
      }

      // --- Success ---
      if (response.ok) {
        return body as T;
      }

      // --- Error mapping ---
      const errorBody = body as ApiErrorBody;
      const code = errorBody?.error?.code ?? "UNKNOWN";
      const message = errorBody?.error?.message ?? response.statusText;

      switch (response.status) {
        case 400:
          throw new ValidationError(message, code);
        case 401:
          throw new AuthenticationError(message, code);
        case 403:
          throw new ForbiddenError(message, code);
        case 404:
          throw new NotFoundError(message, code);
        default:
          if (response.status >= 500) {
            throw new ServerError(message, code, response.status);
          }
          throw new SymbioKBError(
            `Unexpected HTTP ${response.status}: ${message}`,
            code,
            response.status,
          );
      }
    }

    // Should never reach here, but TypeScript requires a return.
    throw new SymbioKBError("Request failed unexpectedly");
  }

  // -----------------------------------------------------------------------
  // Public API — Search
  // -----------------------------------------------------------------------

  /**
   * Full-text search across all pages in the knowledge base.
   *
   * Uses PostgreSQL `websearch_to_tsquery` on the server side, so standard
   * web-search syntax is supported (e.g. `"exact phrase"`, `term1 OR term2`).
   *
   * @param query  The search query string (1-500 characters).
   * @param limit  Maximum number of results (1-100, default 20).
   * @param offset Number of results to skip for pagination (default 0).
   */
  async search(
    query: string,
    options?: { limit?: number; offset?: number },
  ): Promise<SearchResults> {
    const body = await this.request<ApiListResponse<SearchResult>>(
      "GET",
      "/search",
      {
        params: {
          q: query,
          limit: options?.limit ?? 20,
          offset: options?.offset ?? 0,
        },
      },
    );

    return {
      results: body.data,
      total: body.meta.total,
      limit: body.meta.limit,
      offset: body.meta.offset,
    };
  }

  // -----------------------------------------------------------------------
  // Public API — Pages
  // -----------------------------------------------------------------------

  /**
   * Read a single page including its markdown content.
   *
   * @param pageId UUID of the page to read.
   * @throws NotFoundError if the page does not exist.
   */
  async readPage(pageId: string): Promise<Page> {
    const body = await this.request<ApiResponse<Page>>("GET", `/pages/${pageId}`);
    return body.data;
  }

  /**
   * Create a new page in the knowledge base.
   *
   * @param title    Page title (1-255 characters).
   * @param options  Optional markdown content, parent_id, and icon.
   * @throws ValidationError if the title is empty or too long.
   * @throws NotFoundError if the specified parent_id does not exist.
   */
  async createPage(
    title: string,
    options?: {
      markdown?: string;
      parent_id?: string;
      icon?: string;
    },
  ): Promise<Page> {
    const payload: Record<string, unknown> = { title };
    if (options?.markdown !== undefined) payload.markdown = options.markdown;
    if (options?.parent_id !== undefined) payload.parent_id = options.parent_id;
    if (options?.icon !== undefined) payload.icon = options.icon;

    const body = await this.request<ApiResponse<Page>>("POST", "/pages", {
      body: payload,
    });
    return body.data;
  }

  /**
   * Replace the markdown content of an existing page.
   *
   * @param pageId   UUID of the page to update.
   * @param markdown The new markdown content (replaces existing content).
   * @throws NotFoundError if the page does not exist.
   */
  async updatePage(pageId: string, markdown: string): Promise<Page> {
    const body = await this.request<ApiResponse<Page>>("PUT", `/pages/${pageId}`, {
      body: { markdown },
    });
    return body.data;
  }

  /**
   * List pages with optional filtering and pagination.
   *
   * @param options  Pagination, parent filtering, and title search options.
   */
  async listPages(options?: {
    limit?: number;
    offset?: number;
    parent_id?: string;
    search?: string;
  }): Promise<PageListResult> {
    const params: Record<string, string | number> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.parent_id) params.parent_id = options.parent_id;
    if (options?.search) params.search = options.search;

    const body = await this.request<ApiListResponse<Page>>("GET", "/pages", {
      params,
    });

    return {
      pages: body.data,
      total: body.meta.total,
      limit: body.meta.limit,
      offset: body.meta.offset,
    };
  }
}

// ---------------------------------------------------------------------------
// Example usage
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const baseUrl = process.env.SYMBIO_KB_URL ?? "http://localhost:3000";
  const apiKey = process.env.SYMBIO_KB_API_KEY ?? "";

  if (!apiKey) {
    console.error("Set SYMBIO_KB_API_KEY environment variable to run this example.");
    process.exit(1);
  }

  const kb = new SymbioKB({ baseUrl, apiKey });

  // --- List pages ---
  console.log("=== Listing pages ===");
  const pageList = await kb.listPages({ limit: 5 });
  console.log(`Total pages: ${pageList.total}`);
  for (const page of pageList.pages) {
    console.log(`  [${page.id.slice(0, 8)}...] ${page.title}`);
  }

  // --- Create a page ---
  console.log("\n=== Creating a page ===");
  const newPage = await kb.createPage("Agent API Test Page", {
    markdown:
      "# Hello from the TypeScript client\n\nThis page was created via the Agent API.",
    icon: "📘",
  });
  console.log(`Created page: ${newPage.id} — ${newPage.title}`);

  // --- Read the page back ---
  console.log("\n=== Reading page ===");
  const page = await kb.readPage(newPage.id);
  console.log(`Title: ${page.title}`);
  console.log(`Markdown:\n${page.markdown}`);

  // --- Update the page ---
  console.log("\n=== Updating page ===");
  const updated = await kb.updatePage(
    newPage.id,
    "# Updated content\n\nThis content was updated via the Agent API.\n\nSee also: [[Other Page]]",
  );
  console.log(`Updated at: ${updated.updated_at}`);

  // --- Search ---
  console.log("\n=== Searching ===");
  try {
    const results = await kb.search("Agent API");
    console.log(`Found ${results.total} results`);
    for (const r of results.results) {
      console.log(`  [${r.page_id.slice(0, 8)}...] ${r.title} (score: ${r.score.toFixed(3)})`);
      console.log(`    ${r.snippet}`);
    }
  } catch (err) {
    if (err instanceof ValidationError) {
      console.error(`Search validation error: ${err.message}`);
    } else {
      throw err;
    }
  }

  console.log("\nDone.");
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: examples/python/symbio_kb_client.py
"""
SymbioKnowledgeBase Agent API — Python Client

A production-ready client for the SymbioKnowledgeBase Agent API.
Provides typed access to pages, search, and knowledge graph endpoints
with built-in rate limit handling and retry logic.

Usage:
    from symbio_kb_client import SymbioKB

    kb = SymbioKB(base_url="https://kb.example.com", api_key="skb_live_...")
    results = kb.search("deployment guide")
    page = kb.read_page(results[0].page_id)
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Optional

import requests


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class SymbioKBError(Exception):
    """Base exception for all SymbioKB client errors."""

    def __init__(self, message: str, code: Optional[str] = None, status: Optional[int] = None):
        super().__init__(message)
        self.code = code
        self.status = status


class AuthenticationError(SymbioKBError):
    """Raised when the API key is missing, invalid, or revoked (HTTP 401)."""
    pass


class ForbiddenError(SymbioKBError):
    """Raised when the API key lacks the required scope (HTTP 403)."""
    pass


class NotFoundError(SymbioKBError):
    """Raised when the requested resource does not exist (HTTP 404)."""
    pass


class ValidationError(SymbioKBError):
    """Raised when request parameters fail server-side validation (HTTP 400)."""
    pass


class RateLimitError(SymbioKBError):
    """Raised when rate limiting is exceeded and all retries are exhausted (HTTP 429)."""

    def __init__(self, message: str, retry_after: Optional[int] = None, **kwargs):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after


class ServerError(SymbioKBError):
    """Raised for unexpected server-side errors (HTTP 5xx)."""
    pass


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Page:
    """Represents a knowledge base page."""

    id: str
    title: str
    icon: Optional[str] = None
    parent_id: Optional[str] = None
    markdown: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


@dataclass(frozen=True)
class SearchResult:
    """A single result from a full-text search query."""

    page_id: str
    title: str
    snippet: str
    score: float
    icon: Optional[str] = None


@dataclass(frozen=True)
class PageList:
    """Paginated list of pages returned by list_pages."""

    pages: list[Page] = field(default_factory=list)
    total: int = 0
    limit: int = 50
    offset: int = 0


@dataclass(frozen=True)
class SearchResults:
    """Paginated list of search results returned by search."""

    results: list[SearchResult] = field(default_factory=list)
    total: int = 0
    limit: int = 20
    offset: int = 0


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------


class SymbioKB:
    """Client for the SymbioKnowledgeBase Agent API.

    Args:
        base_url: The base URL of the SymbioKnowledgeBase instance
                  (e.g. ``"https://kb.example.com"``).  Trailing slashes are
                  stripped automatically.
        api_key:  A bearer token.  For API key auth use a key in the format
                  ``skb_live_<hex>``.  For Supabase JWT auth pass the JWT
                  string directly.
        timeout:  HTTP request timeout in seconds.  Defaults to 30.
        max_retries: Maximum number of automatic retries on 429 responses.
                     Defaults to 3.
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        timeout: int = 30,
        max_retries: int = 3,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries

        self._session = requests.Session()
        self._session.headers.update(
            {
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        )

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _url(self, path: str) -> str:
        """Build a full URL for the given API path."""
        return f"{self.base_url}/api/agent{path}"

    def _request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[dict] = None,
        json_body: Optional[dict] = None,
    ) -> dict:
        """Execute an HTTP request with rate-limit retry logic.

        Returns the parsed JSON response body on success.
        Raises a typed exception on any error.
        """
        url = self._url(path)
        last_exception: Optional[Exception] = None

        for attempt in range(self.max_retries + 1):
            try:
                resp = self._session.request(
                    method,
                    url,
                    params=params,
                    json=json_body,
                    timeout=self.timeout,
                )
            except requests.RequestException as exc:
                raise SymbioKBError(f"HTTP request failed: {exc}") from exc

            # --- Rate limit handling (429) ---
            if resp.status_code == 429:
                retry_after = int(resp.headers.get("Retry-After", "5"))
                if attempt < self.max_retries:
                    time.sleep(retry_after)
                    continue
                last_exception = RateLimitError(
                    "Rate limit exceeded after all retries",
                    retry_after=retry_after,
                    code="RATE_LIMIT_EXCEEDED",
                    status=429,
                )
                break

            # --- Parse response body ---
            try:
                body = resp.json()
            except ValueError:
                raise ServerError(
                    f"Non-JSON response (HTTP {resp.status_code})",
                    status=resp.status_code,
                )

            # --- Success ---
            if 200 <= resp.status_code < 300:
                return body

            # --- Error mapping ---
            error_obj = body.get("error", {})
            code = error_obj.get("code", "UNKNOWN")
            message = error_obj.get("message", resp.text)

            if resp.status_code == 401:
                raise AuthenticationError(message, code=code, status=401)
            if resp.status_code == 403:
                raise ForbiddenError(message, code=code, status=403)
            if resp.status_code == 404:
                raise NotFoundError(message, code=code, status=404)
            if resp.status_code == 400:
                raise ValidationError(message, code=code, status=400)
            if resp.status_code >= 500:
                raise ServerError(message, code=code, status=resp.status_code)

            raise SymbioKBError(
                f"Unexpected HTTP {resp.status_code}: {message}",
                code=code,
                status=resp.status_code,
            )

        if last_exception:
            raise last_exception

        # Should never reach here, but satisfy the type checker.
        raise SymbioKBError("Request failed unexpectedly")  # pragma: no cover

    # ------------------------------------------------------------------
    # Public API — Search
    # ------------------------------------------------------------------

    def search(
        self,
        query: str,
        *,
        limit: int = 20,
        offset: int = 0,
    ) -> SearchResults:
        """Full-text search across all pages in the knowledge base.

        Uses PostgreSQL ``websearch_to_tsquery`` on the server side, so
        standard web-search syntax is supported (e.g. ``"exact phrase"``,
        ``term1 OR term2``).

        Args:
            query:  The search query string (1-500 characters).
            limit:  Maximum number of results to return (1-100, default 20).
            offset: Number of results to skip for pagination (default 0).

        Returns:
            A ``SearchResults`` object containing matching pages with
            context snippets and relevance scores.

        Raises:
            ValidationError: If query parameters are invalid.
        """
        body = self._request(
            "GET",
            "/search",
            params={"q": query, "limit": limit, "offset": offset},
        )
        data = body.get("data", [])
        meta = body.get("meta", {})
        return SearchResults(
            results=[
                SearchResult(
                    page_id=r["page_id"],
                    title=r["title"],
                    snippet=r["snippet"],
                    score=r["score"],
                    icon=r.get("icon"),
                )
                for r in data
            ],
            total=meta.get("total", len(data)),
            limit=meta.get("limit", limit),
            offset=meta.get("offset", offset),
        )

    # ------------------------------------------------------------------
    # Public API — Pages
    # ------------------------------------------------------------------

    def read_page(self, page_id: str) -> Page:
        """Read a single page including its markdown content.

        Args:
            page_id: UUID of the page to read.

        Returns:
            A ``Page`` object with the ``markdown`` field populated.

        Raises:
            NotFoundError: If the page does not exist.
        """
        body = self._request("GET", f"/pages/{page_id}")
        d = body.get("data", {})
        return Page(
            id=d["id"],
            title=d["title"],
            icon=d.get("icon"),
            parent_id=d.get("parent_id"),
            markdown=d.get("markdown", ""),
            created_at=d.get("created_at"),
            updated_at=d.get("updated_at"),
        )

    def create_page(
        self,
        title: str,
        *,
        markdown: Optional[str] = None,
        parent_id: Optional[str] = None,
        icon: Optional[str] = None,
    ) -> Page:
        """Create a new page in the knowledge base.

        Args:
            title:     Page title (1-255 characters).
            markdown:  Optional markdown content for the page body.
            parent_id: Optional UUID of a parent page for nesting.
            icon:      Optional emoji or icon string.

        Returns:
            A ``Page`` object for the newly created page.

        Raises:
            ValidationError: If the title is empty or too long.
            NotFoundError: If the specified parent_id does not exist.
        """
        payload: dict = {"title": title}
        if markdown is not None:
            payload["markdown"] = markdown
        if parent_id is not None:
            payload["parent_id"] = parent_id
        if icon is not None:
            payload["icon"] = icon

        body = self._request("POST", "/pages", json_body=payload)
        d = body.get("data", {})
        return Page(
            id=d["id"],
            title=d["title"],
            created_at=d.get("created_at"),
        )

    def update_page(self, page_id: str, markdown: str) -> Page:
        """Replace the markdown content of an existing page.

        Args:
            page_id:  UUID of the page to update.
            markdown: The new markdown content (replaces existing content).

        Returns:
            A ``Page`` object with the updated timestamp.

        Raises:
            NotFoundError: If the page does not exist.
            ValidationError: If the markdown field is missing.
        """
        body = self._request(
            "PUT",
            f"/pages/{page_id}",
            json_body={"markdown": markdown},
        )
        d = body.get("data", {})
        return Page(
            id=d["id"],
            title=d.get("title", ""),
            updated_at=d.get("updated_at"),
        )

    def list_pages(
        self,
        *,
        limit: int = 50,
        offset: int = 0,
        parent_id: Optional[str] = None,
        search: Optional[str] = None,
    ) -> PageList:
        """List pages with optional filtering and pagination.

        Args:
            limit:     Maximum number of pages to return (1-100, default 50).
            offset:    Number of pages to skip for pagination (default 0).
            parent_id: Filter to only children of this parent page UUID.
            search:    Filter pages whose title contains this string
                       (case-insensitive).

        Returns:
            A ``PageList`` object with pagination metadata.
        """
        params: dict = {"limit": limit, "offset": offset}
        if parent_id is not None:
            params["parent_id"] = parent_id
        if search is not None:
            params["search"] = search

        body = self._request("GET", "/pages", params=params)
        data = body.get("data", [])
        meta = body.get("meta", {})
        return PageList(
            pages=[
                Page(
                    id=p["id"],
                    title=p["title"],
                    icon=p.get("icon"),
                    parent_id=p.get("parent_id"),
                    created_at=p.get("created_at"),
                    updated_at=p.get("updated_at"),
                )
                for p in data
            ],
            total=meta.get("total", len(data)),
            limit=meta.get("limit", limit),
            offset=meta.get("offset", offset),
        )

    # ------------------------------------------------------------------
    # Context manager support
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Close the underlying HTTP session."""
        self._session.close()

    def __enter__(self) -> SymbioKB:
        return self

    def __exit__(self, *exc) -> None:
        self.close()


# ---------------------------------------------------------------------------
# Example usage
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import os
    import sys

    base_url = os.environ.get("SYMBIO_KB_URL", "http://localhost:3000")
    api_key = os.environ.get("SYMBIO_KB_API_KEY", "")

    if not api_key:
        print("Set SYMBIO_KB_API_KEY environment variable to run this example.")
        sys.exit(1)

    with SymbioKB(base_url=base_url, api_key=api_key) as kb:
        # --- List pages ---
        print("=== Listing pages ===")
        page_list = kb.list_pages(limit=5)
        print(f"Total pages: {page_list.total}")
        for page in page_list.pages:
            print(f"  [{page.id[:8]}...] {page.title}")

        # --- Create a page ---
        print("\n=== Creating a page ===")
        new_page = kb.create_page(
            title="Agent API Test Page",
            markdown="# Hello from the Python client\n\nThis page was created via the Agent API.",
            icon="🐍",
        )
        print(f"Created page: {new_page.id} — {new_page.title}")

        # --- Read the page back ---
        print("\n=== Reading page ===")
        page = kb.read_page(new_page.id)
        print(f"Title: {page.title}")
        print(f"Markdown:\n{page.markdown}")

        # --- Update the page ---
        print("\n=== Updating page ===")
        updated = kb.update_page(
            new_page.id,
            markdown="# Updated content\n\nThis content was updated via the Agent API.\n\nSee also: [[Other Page]]",
        )
        print(f"Updated at: {updated.updated_at}")

        # --- Search ---
        print("\n=== Searching ===")
        try:
            results = kb.search("Agent API")
            print(f"Found {results.total} results")
            for r in results.results:
                print(f"  [{r.page_id[:8]}...] {r.title} (score: {r.score:.3f})")
                print(f"    {r.snippet}")
        except ValidationError as exc:
            print(f"Search validation error: {exc}")

    print("\nDone.")
